Add PVC-to-Pod cache to improve volume policy performance

shubham-pampattiwar opened the pull request:
Thank you for contributing to Velero!

# Please add a summary of your change

The `GetPodsUsingPVC` function had O(N*M) complexity - for each PVC (N), it listed ALL pods in the namespace and iterated through each pod (M). With many PVCs and pods in a cluster, this caused significant performance degradation during backup operations.

As reported in issue #9169, one N*M iteration even took 2 seconds:
```
time="2025-08-08T04:03:22Z" level=info msg="Executing takePVSnapshot" ...
time="2025-08-08T04:03:24Z" level=info msg="skipping snapshot action for pv pvc-e74b7300..."
```

This change introduces a **PVC-to-Pod cache** that is built once per backup and reused for all PVC lookups, reducing complexity from O(N*M) to O(N+M).

**Changes:**
- Add `PVCPodCache` struct with thread-safe caching in `pkg/util/podvolume`
- Add `NewVolumeHelperImplWithCache` constructor for cache support  
- Build cache before backup item processing in `backup.go`
- Add comprehensive unit tests for cache functionality
- Graceful fallback to direct lookups if cache building fails

**Performance Impact:**
| Scenario | Before | After |
|----------|--------|-------|
| 100 PVCs, 1000 pods | ~100,000 iterations | ~1,100 iterations |
| Cache lookup | O(N*M) per PVC | O(1) per PVC |

# Does your change fix a particular issue?

Fixes #9179

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

kaovilai commented on pkg/backup/backup.go:
```diff
@@ -408,6 +409,21 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
+	// This avoids O(N*M) complexity when there are many PVCs and pods.
+	// See issue #9179 for details.
+	pvcPodCache := podvolumeutil.NewPVCPodCache()
+	namespaces := backupRequest.NamespaceIncludesExcludes.GetIncludes()
```
```suggestion
	namespaces := backupRequest.NamespaceIncludesExcludes.ResolveNamespaceList()
```

Lyndon-Li commented on pkg/backup/backup.go:
```diff
@@ -408,6 +409,24 @@ func (kb *kubernetesBackupper) BackupWithResolvers(
+	// Build PVC-to-Pod cache for improved volume policy lookup performance.
+	// This avoids O(N*M) complexity when there are many PVCs and pods.
+	// See issue #9179 for details.
+	pvcPodCache := podvolumeutil.NewPVCPodCache()
```
Since this `pvcPodCache` is used by `volumehelper` only, why don't we move this cache build into volumehelper? Benefits:
1. All the logics are in `volumehelper`, no need to change the backupper
2. `volumehelper` may be per backup, or it is also reasonable per backups which is started in the same period of time, having the cache in `volumehelper` ensures the cache always have the same lifecycle with the `volumehelper`

Lyndon-Li commented on pkg/util/podvolume/pod_volume.go:
```diff
@@ -165,6 +284,27 @@ func GetPodsUsingPVC(
+	}
+
+	// Fall back to direct lookup
+	return GetPodsUsingPVC(pvcNamespace, pvcName, crClient)
```
See https://github.com/vmware-tanzu/velero/pull/9441/files#r2609213480

shubham-pampattiwar commented:
@Lyndon-Li, thanks for the review feedback. I've addressed your comments:

1. Moved cache building into volumehelper - Renamed `NewVolumeHelperImplWithCache` to `NewVolumeHelperImplWithNamespaces`. It now takes a list of namespaces and builds the cache internally, keeping all the logic encapsulated in volumehelper.
2. Removed fallback in main backup path - `NewVolumeHelperImplWithNamespaces` now returns (VolumeHelper, error). If cache build fails, backup fails with an error instead of silently falling back to direct lookups. This ensures predictable performance.
3. Code reuse - `NewVolumeHelperImpl` (used by plugins) now calls `NewVolumeHelperImplWithNamespaces` with nil namespaces.

The fallback logic still exists in `GetPodsUsingPVCWithCache` but it's only triggered for plugins that don't use the cache. In the main backup path, the cache is always built or the backup fails.

Lyndon-Li commented:
@shubham-pampattiwar 

I realized we cannot leave plugins to call the original `NewVolumeHelperImpl`, either based on my suggestion or your original change:
1. `NewVolumeHelperImpl` is not only called in the backupper, but also called by plugins. E.g., for BIA of each PVC
2. With my suggestion, `NewVolumeHelperImpl` is called by each BIA, and cache will be built for multiple times
3. With your original change, `NewVolumeHelperImpl` is called by each BIA, since it fallbacks to the original behavior, so for each call, pod-pvc list is created once

We need a further consideration. With the current PR, I am afraid we cannot solve the original problem.

shubham-pampattiwar commented:
@Lyndon-Li You are right that the CSI PVC BIA (pvc_action.go) still calls `ShouldPerformSnapshotWithBackup` which creates a new `VolumeHelper` each time.

Looking at the code paths:

1. Main backup path (`item_backupper.go`) - Uses `ib.volumeHelperImpl.ShouldPerformSnapshot` which now has the cache. This should be fixed with the current PR.
2. CSI PVC BIA (`pvc_action.go`) - Uses `ShouldPerformSnapshotWithBackup` which creates a new `VolumeHelper` per call. This is still O(N*M).

To fix the BIA path, I see a couple of options:

Option A: Backup-level cache registry
Store the cache in a package-level map keyed by backup name. The backupper registers the cache at backup start, and BIAs look it up by backup name. Cleanup happens at backup end.

Option B: Modify CSI PVC action directly
Since it's a built-in BIA, we could have it build/hold its own cache. This depends on whether the BIA instance is reused across all PVCs in a backup or recreated each time.

What approach would you prefer? Or should we scope this PR to the main backup path fix and address the BIA issue in a follow-up? Let me know what you think ?

Lyndon-Li commented:
> @Lyndon-Li Let me clarify a few things:
> 
> Option A: Backup-level cache registry
> 
> * Create a new package (e.g., `pkg/backup/cache`) with a global registry
> * Registry uses `sync.RWMutex `protected `map: map[string]*PVCPodCache`
> * Main backup path builds cache once and calls `cache.RegisterCacheForBackup(backup.UID, pvcPodCache)`
> * BIAs import the package and call `cache.GetCacheForBackup(backup.UID)` to retrieve the cache
> * `ShouldPerformSnapshotWithBackup` uses the retrieved cache for lookups
> * Cleanup via `cache.UnregisterCacheForBackup(backup.UID)` when backup finishes
> 
> Option B: Per-backup cache tracking in BIA
> 
> * Similar to https://github.com/openshift/openshift-velero-plugin/blob/oadp-dev/velero-plugins/serviceaccount/backup.go#L26
> * The BIA struct holds a cache with `UpdatedForBackup map[string]bool`
> * On first `Execute()` for a backup, BIA builds the cache itself and marks it updated
> * Subsequent `Execute()` calls within that backup reuse the cached data

For Option A: Just notice that BIAs and Velero backupper are running in different **PROCESSES**, so there is no way to share the cache data without the inter-process data sharing methods, e.g., protocolBuffer, share memory, etc.
For Option B: Notice that BIAs are host by plugin processes which are not long running processes. The plugin framework may stop the process after ONLY ONE BIA call and then start another process later. So the cache in the plugin processes won't work as the expectation.

Lyndon-Li commented:
@shubham-pampattiwar @sseago 
I think the fact is:
A. The current cache solution could fix the problem for the calls happening in backup
B. The current cache solution could also fix the problem in plugins more or less if we have a per plugin process cache
C. The current cache solution is not good enough for plugin since the behavior is not expectable --- if multiple calls are executed in the same plugin process, it works; otherwise, it doesn't work

If you agree on the facts, I suggest:
We modify the current PR to make a per process cache for plugins:
1. The original way which initialize a cache out of `VolumeHelperImpl` could work
2. But I think it is better to have the cache inside `VolumeHelperImpl` and have a global `VolumeHelperImpl` instead --- it is not necessary to expose the cache details to the callers
3. If you want to go with 2, based on the current PR, you could call `NewVolumeHelperImpl` in a global place in the plugin instead of calling it multiple times in `ShouldPerformXXX` functions

After we make this modification, the PR would meet to A and B.
For C, let's open another issue to track it and seek for a ultimate solution.

Lyndon-Li commented on pkg/plugin/utils/volumehelper/volume_policy_helper.go:
```diff
@@ -60,3 +92,92 @@ func ShouldPerformSnapshotWithBackup(
+// If namespaces is nil or empty, the function will resolve the namespace list from the backup spec.
+// If backup.Spec.IncludedNamespaces is empty (meaning all namespaces), it will list all namespaces
+// from the cluster.
+func NewVolumeHelperForBackup(
```
I don't see this function is used, can we remove it if not?
From the comments of this function, it seems to be used by plugins. However, plugins don't have enough information to provide the correct namespace list in the backup, the backup may have multiple ways to opt-in/opt-out namespaces. Regarding to this, we should not provide a function like this to plugins.

Lyndon-Li commented on pkg/plugin/utils/volumehelper/volume_policy_helper.go:
```diff
@@ -40,6 +43,35 @@ func ShouldPerformSnapshotWithBackup(
+// avoiding the overhead of creating a new VolumeHelper on each call.
+// This is useful for BIA plugins that process multiple PVCs during a single backup
+// and want to reuse the same VolumeHelper (with its internal cache) across calls.
+func ShouldPerformSnapshotWithVolumeHelper(
```
Since this function is called `XXXWithVolumeHelper`, we don't need to consider the case that `vh == nil`, we can return error. Otherwise,  it overlaps with `ShouldPerformSnapshotWithBackup`

Lyndon-Li commented on pkg/util/podvolume/pod_volume.go:
```diff
@@ -109,12 +253,46 @@ func GetVolumesToExclude(obj metav1.Object) []string {
 
+// IsPVCDefaultToFSBackup checks if a PVC should default to fs-backup based on pod annotations.
+// Deprecated: Use IsPVCDefaultToFSBackupWithCache for better performance.
 func IsPVCDefaultToFSBackup(pvcNamespace, pvcName string, crClient crclient.Client, defaultVolumesToFsBackup bool) (bool, error) {
```
If this one is not used by plugins and it is deprecated, why don't we remove it directly?

Lyndon-Li commented on pkg/util/podvolume/pod_volume.go:
```diff
@@ -140,6 +318,9 @@ func getPodVolumeNameForPVC(pod corev1api.Pod, pvcName string) (string, error) {
 
+// GetPodsUsingPVC returns all pods in the given namespace that use the specified PVC.
+// This function lists all pods in the namespace and filters them, which can be slow
+// when there are many pods. Consider using GetPodsUsingPVCWithCache for better performance.
```
If this one is not used by plugins and it is deprecated, why don't we remove it directly?
